File: lib/ionicons/font_generator.rb
require 'mkmf'

module Ionicons
  class FontGenerator
    const :tools, %w{fontforge sfnt2woff}
    const :fonts_path, Rails.root.join('frontend/fonts')
    const :styles_path, Rails.root.join('frontend/styles')
    const :repo_path, Pathname.new(__dir__).join('ionicons')
    const :repo_url, 'git@github.com:driftyco/ionicons.git'

    def call
      check_tools
      with_repo do
        remove_unused_svgs
        build_font
        copy_font
      end
    end

    private

    def check_tools
      missing = tools.reject { |tool| find_executable tool }
      return if missing.empty?
      raise StandardError.new("One or more of the required tools are missing: #{missing.join(', ')}. Install them first")
    end

    def remove_unused_svgs
      Dir.glob(repo_path.join('src/*.svg'))
        .reject { |path| icons.include? path.split('/').last.chomp('.svg') }
        .each { |path| File.delete path }
    end

    def build_font
      system "python #{repo_path.join('builder/generate.py')}"
    end

    def copy_font
      FileUtils.cp repo_path.join('fonts/ionicons.woff'), fonts_path
    end

    def icons
      @icons ||= begin
        direct_icons = styles.lines
          .select { |line| line.include? '+ion-icon' }
          .map { |line| line.match(/ion-icon\((?:(?:'|"))?(.*?)(?:(?:'|"))?\)/)[1] }
          .reject { |icon| icon.starts_with? '$' }

        each_icons = styles.lines
          .select { |line| line.include? '@each' }
          .flat_map { |line| EachLine.new(line).icons }

        (direct_icons + each_icons).to_set
      end
    end

    def styles
      @styles ||= Dir.glob(styles_path.join('**/*.sass')).map{ |path| File.read path }.join("\n")
    end

    def with_repo
      clone_repo
      yield
      remove_repo
    end

    def clone_repo
      system "git clone #{repo_url} #{repo_path}"
    end

    def remove_repo
      FileUtils.rm_r repo_path
    end
  end
end

File: lib/ionicons/font_generator.exs
Code.require_file "each_line.exs", __DIR__

defmodule Ionicons.FontGenerator do
  alias Ionicons.EachLine

  @tools ~W{fontforge sfnt2woff}
  @font_path "frontend/fonts/ionicons.woff"
  @styles_path "frontend/styles"
  @repo_path Path.join(__DIR__, "ionicons")
  @repo_url "git@github.com:driftyco/ionicons.git"
  @direct_icon_regex ~r/ion-icon\((?:(?:'|"))?(.*?)(?:(?:'|"))?\)/

  def generate do
    check_tools
    with_repo fn ->
      remove_unused_svgs(icons)
      build_font
      copy_font
    end
    :ok
  end

  defp check_tools do
    missing = Enum.reject @tools, &System.find_executable/1
    unless Enum.empty? missing do
      raise RuntimeError, message: "One or more of the required tools are missing: #{Enum.join missing, ", "}. Install them first"
    end
  end

  defp remove_unused_svgs(icons) do
    @repo_path
    |> Path.join("src/**/*.svg")
    |> Path.wildcard
    |> Enum.reject(&use_svg?(&1, icons))
    |> Enum.each(&File.rm!/1)
  end

  defp build_font do
    System.cmd "python", [Path.join(@repo_path, "builder/generate.py")]
  end

  defp copy_font do
    @repo_path
    |> Path.join("fonts/ionicons.woff")
    |> File.cp!(@font_path)
  end

  defp use_svg?(path, icons) do
    Set.member? icons, Path.basename(path, ".svg")
  end

  defp icons do
    lines = String.split styles, "\n"
    direct_icons(lines) ++ each_icons(lines)
    |> Enum.reduce(MapSet.new, &Set.put(&2, &1))
  end

  defp direct_icons(lines) do
    lines
    |> Enum.filter(&String.contains?(&1, "+ion-icon"))
    |> Enum.map(&Regex.run @direct_icon_regex, &1, capture: :all_but_first)
    |> Enum.map(&List.first/1)
    |> Enum.reject(&String.starts_with?(&1, "$"))
  end

  defp each_icons(lines) do
    lines
    |> Enum.filter(&String.contains?(&1, "@each"))
    |> Enum.flat_map(&EachLine.icons(&1))
  end

  defp styles do
    @styles_path
    |> Path.join("**/*.sass")
    |> Path.wildcard
    |> Enum.map(&File.read!/1)
    |> Enum.join("\n")
  end

  defp with_repo(fun) do
    clone_repo
    fun.()
    remove_repo
  end

  defp clone_repo do
    System.cmd "git", ~w{clone #{@repo_url} #{@repo_path}}
  end

  defp remove_repo do
    File.rm_rf! @repo_path
  end
end
